sympy/physics/quantum/qapply.py

```diff
@@ -1,206 +1,757 @@
+          ``TensorProduct``s (same option as in ``expand``) (default: value of option ``mul``).
+        * ``power_base``: for commutative factors, split powers of multiplied bases
+          (same option as in ``expand``) (default: value of option ``mul``).
+        * ``power_exp``: for cummutative factors, expand addition in exponents into
+          multiplied bases (same option as in ``expand``) (default: False).
+
+
+
+    Returns
+    =======
+
+    e : Expr
+        The original expression, but with the operators applied to operators and states
+        and with moderate expansion done by distributing commutative factors over summands.
+
+    Examples
+    ========
+
+        >>> from sympy.physics.quantum import qapply, Ket, Bra
+        >>> from sympy.physics.quantum.qubit import QubitBra
+        >>> from sympy.physics.quantum.gate import XGate
+        >>> from sympy.core.symbol import symbols
+        >>> b = Bra('b')
+        >>> k = Ket('k')
+        >>> A = k * b
+        >>> A
+        |k><b|
+        >>> qapply(A * b.dual / (b * b.dual))
+        |k>
+        >>> qapply(k.dual * A / (k.dual * k))
+        <b|
+        >>> qapply(QubitBra(1) * XGate(0), dagger = True)
+        <0|
+        >>> qapply(A * k * b * A, op_join = False)
+        <b|k>**2*|k>*<b|
+        >>> n = symbols("n", integer=True, nonnegative=True)
+        >>> qapply(A ** (n + 2))
+        <b|k>**(n + 1)*|k><b|
+
+    """
+    # moved import here to avoid circular reference on inport qapply->Density->represent->qapply
+    from sympy.physics.quantum.density import Density
+
+    # Note: The variables EmptyMul, options, dagger, ip_doit, op_join
+    # are assigned and in read-only scope: see bottom of file for body of qapply()
+
+    """
+    Remarks on quantum type constructors and Mul
+    --------------------------------------------
+
+    1.) The code relies on the auto-simplifications that Mul and all the composite objects constructors like
+    InnerProduct, OuterProduct, TensorProduct, Pow etc do: Extraction of commutative factors, contracting
+    of same factors to Powers etc, so these constructors may return their type, Mul or even other types.
+
+    2.) Mul(a,b) != a * b:
+    Mul(a, b, c) processes the arguments a, b, c (e.g. factoring out scalars, grouping same factors to powers)
+    but does not try to actually multiply the arguments, i.e. is not trying (a*(b*c)), (a*b) etc.
+    On the other hand a*b is executed using *-operator __mul__ for type(a), type(b):
+    E.g. for Kets and Bras, __mul__ is overloaded to form InnerProducts resp. OuterProducts types, and
+    e.g. IdentityOperator()*Ket(1)-> Ket(1)
+    (Examples: Mul(Bra(1), Ket(1)) returns unchanged. Bra(1)*Mul(Ket(1), Bra(1)) returns
+    Mul(Bra(1), Ket(1), Bra(1)), while Bra(1)*Ket(1)*Bra(1) returns Mul(InnerProduct(Bra(1),Ket(1)), Bra(1)).
+    So while the return type of Mul(a,b) will be Mul (or Zero) the return type of a*b completely depends
+    on the input types.

```

Comment:
I realise this is not for this PR but ideally there would be no distinction between `a*b` and `Mul(a, b)`. In general it is better not to overload `__mul__` in `Expr` subclasses but rather to find a way of making `Mul` do the right thing.

Reply:
Matrices handle this using `Mul._kind_dispatcher`:
https://github.com/sympy/sympy/blob/40b1af74c7f6826b9ba702c359cb538621e9207c/sympy/matrices/matrices.py#L751-L775
That is intended to be a better approach for this sort of thing but I don't know if the design of it would work for what `qapply` does. It requires defining the `.kind` attributes on quantum objects.